fix(#9799): add user-agent header to outbound

witash opened the pull request:
adds user-agent header to outbound
User-Agent headers should include versions, which means a little extra stuff; it tries to get the CHT version in the same way as api/src/services/deploy-info.js, but that code is not directly available from shared-libs, so there is some small duplication. 
Added getVersion() to shared-libs/settings. It doesn't quite fit there, but I don't think its worth it to make another separate shared-lib

We could also just omit the version if it adds too much mess, but I think these small additions are ok.

closes #9799

dianabarsan reviewed (commented):
Such a quick turnaround! nice! 
I left some minor suggestions inline.

gitcliff commented:
hello @dianabarsan @witash should we upgrade to this branch in the UI in the meantime as the work here still goes on ?

witash commented:
@dianabarsan this is ready for re-reivew, moved deploy-info to shared-libs/environment so it could be used in both the API and other shared-libs, and refactored outbound.js to use aysnc/await instead of explicit Promises.

@gitcliff I think it should be ok to upgrade your instance to this branch temporarily

dianabarsan commented on shared-libs/environment/src/index.js:
```diff
@@ -57,5 +63,48 @@ try {
+  try {
+    ddoc = await request.get({
+      url: `${couchUrl}/_design/${module.exports.ddoc}`,
+      json: true
```
This should not be required (it should be enabled by default).

dianabarsan commented on shared-libs/environment/src/index.js:
```diff
@@ -57,5 +63,48 @@ try {
+
+  let ddoc;
+  try {
+    ddoc = await request.get({
```
I believe `ddoc` can be a scoped `const` here, and we can return within the try-catch block.

dianabarsan commented on shared-libs/environment/test/index.spec.js:
```diff
@@ -38,7 +39,7 @@ describe('environment', () => {
     const env = rewire('../src/index');
 
-    expect(env).to.deep.equal({
+    expect(env).to.deep.include({
```
I think we should keep the strict `equal` test here, and be definitive of which properties should be exported, and we can use `chai-exclude` to exclude the functions.

dianabarsan commented on shared-libs/outbound/src/outbound.js:
```diff
@@ -92,101 +95,118 @@ const mapDocumentToPayload = (doc, config, key) => {
+  const sendOptions = {
+    url: urlJoin(config.destination.base_url, config.destination.path),
+    body: payload,
+    json: true,
```
`json` is `true` by default.

dianabarsan commented on shared-libs/outbound/src/outbound.js:
```diff
@@ -92,101 +95,118 @@ const mapDocumentToPayload = (doc, config, key) => {
-        clone.auth.password = '*****';
-      }
-      logger.debug(JSON.stringify(clone, null, 2));
+  sendOptions.headers['user-agent'] = await getUserAgent();
```
Why not add this directly when declaring `sendOptions`?

dianabarsan reviewed (commented):
Very nice. Just some minor comments here.

gitcliff commented on shared-libs/outbound/src/outbound.js:
```diff
@@ -16,22 +16,25 @@ const urlJoin = require('url-join');
+  if (!password) {
+    throw new OutboundError(
+      `Credentials for '${key}' have not been configured. See the Outbound documentation.`
+    );
```
hello @witash been testing this out and what is the sample payload for this new refactoring i am hitting this error message with an outbound payload like 
```
 "patient triggers enroll workflow in Textit": {
      "relevant_to": "doc.type === 'person' && doc.docket",
      "destination": {
        "base_url": "https://textit.com",
        "auth": {
          "type": "header",
          "name": "Authorization",
          "value_key": "rapidpro.api"
        },
        "headers": {
          "User-Agent": "CHT/4.17.0"
        },
        "path": "/api/v2/flow_starts.json"
      },
      "mapping": {
        "flow": {
          "expr": "'650567d6-6662-4c0e-9ea3-98eba8ef'"
        },
        "urns": {
          "expr": "['tel:' + doc.phone]"
        }
      }
    },

```
Error log `2025-03-03T20:28:13.381 ERROR: Failed to push 356ffef1-29d8-44a7-8861-dc5d3637 to patient triggers enroll workflow in Textit: Credentials for 'rapidro.api' have not been configured. See the Outbound documentation. 
`

andrablaj commented:
Per our discussion today, @witash, re-requesting a review from @dianabarsan

dianabarsan commented on shared-libs/environment/test/index.spec.js:
```diff
@@ -74,4 +80,98 @@ describe('environment', () => {
+          timestamp: '2023-01-02'
+        });
+        expect(request.get.callCount).to.equal(1);
+        expect(request.get.args[0][0].url).to.equal('http://admin:pass@localhost:5984/medicdb/_design/medic');
```
I think we should be more strict about the assertion here, and assert all params that are passed to request.get. 
We could easily make a `DELETE` request and the test would still pass

dianabarsan commented on shared-libs/environment/test/index.spec.js:
```diff
@@ -74,4 +80,98 @@ describe('environment', () => {
+
+        const env = rewire('../src/index');
+
+        try {
```
A nicer way to assert on calls we know will fail is do something like: 
```suggestion
await expect(env.getDeployInfo()).to.eventually.be.rejectedWith('Failed to fetch');
```

This requires `chai.use(require('chai-as-promised''))`; you should be able to find examples of usage throughout our codebase.

dianabarsan commented on shared-libs/outbound/test/outbound.spec.js:
```diff
@@ -340,10 +341,7 @@ describe('outbound shared library', () => {
           assert.equal(request.post.args[0][0].url, 'http://test/foo');
           assert.deepEqual(request.post.args[0][0].body, {some: 'data'});
-          assert.equal(request.post.args[0][0].json, true);
-          assert.deepEqual(request.post.args[0][0].headers, {
```
I generally prefer to keep test assertions with as much parameter coverage as possible, So here, instead of reducing coverage, can you please update the test to assert other potential headers too instead of reducing the assertion to just one header?

dianabarsan commented on shared-libs/outbound/test/outbound.spec.js:
```diff
@@ -267,14 +268,15 @@ describe('outbound shared library', () => {
           assert.equal(request.post.args[0][0].url, 'http://test/foo');
           assert.deepEqual(request.post.args[0][0].body, {some: 'data'});
-          assert.equal(request.post.args[0][0].json, true);
+          assert.match(request.post.args[0][0].headers['user-agent'], /^CHT\/4\.18\.0/);
```
Can you please add a test where os.platform and arch are also stubbed, so we see the complete user agent header as well?

dianabarsan reviewed (commented):
Very nice! Just a few minor requests inline :) Thanks for your patience.

dianabarsan reviewed (approved):
Approving to unblock this. Good luck!!
